libraries/AP_Scripting/AP_Scripting_CANSensor.h

```diff
@@ -0,0 +1,43 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  Scripting CANSensor class, for easy scripting CAN support
+ */
+ 
+#pragma once
+
+#include <AP_CANManager/AP_CANSensor.h>
+
+class ScriptingCANSensor : public CANSensor {
+public:
+
+    ScriptingCANSensor(AP_CANManager::Driver_Type type, uint32_t buffer_size):CANSensor("Script", type),buffer(buffer_size) {};
+
+    // handler for outgoing frames, using uint32
+    bool write_frame(AP_HAL::CANFrame &out_frame, const uint32_t timeout_us) {
+        return CANSensor::write_frame(out_frame, timeout_us);
+    };
+
+    // handler for incoming frames, add to buffer
+    void handle_frame(AP_HAL::CANFrame &frame) override { buffer.push(frame); }
+
+    // read a frame from the buffer
+    bool read_frame(AP_HAL::CANFrame &frame) { return buffer.pop(frame); }
```

Comment:
you need semaphore locking here, CANSensor runs on a thread.

Reply:
I'm using the thread safe ObjectBuffer, it does its own locking.

Reply:
I assume that is sufficient anyway, its still locking up AP completely if there are messages before the script has run.